src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIStreams.ValueTask.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using System;
+
+namespace Microsoft.Data.SqlClient.SNI
+{
+    internal sealed partial class SNISslStream
+    {
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ValueTask<int> valueTask = ReadAsync(new Memory<byte>(buffer, offset, count), cancellationToken);
+            if (valueTask.IsCompletedSuccessfully)
+            {
+                return Task.FromResult(valueTask.Result);
+            }
+            else
+            {
+                return valueTask.AsTask();
```

Comment:
What are the benefits of starting with a value task and returning a task? can we just start with a task and return a task?

Reply:
The only version that does the real work is the ValueTask version so everything must call that one and then convert the result to whatever upstream type it needs. I agree that it looks wasteful but it's more complex to have two complete implementations and this is the pattern used in /runtime 

In an ideal world all callers will use ValueTask directly and avoid the Task. In cases where somehow the caller can't do that (isn't aware of ValueTask) they'll incur the same task allocation penalty they were already going to have to and the only change is a couple of additional stack copy instructions. Overall it's a net win. 

Since we (sqlclient) are the only caller of these functions this PR completes the vertical call chain that uses ValueTask instead of Task and we'll see the benefits, this method might only get called through sequential reader streams.

Reply:
In several cases like this, the consumer is not waiting them directly they are waiting for a Task and that makes me wonder if value task is a good choice.


> In an ideal world all callers will use ValueTask directly and avoid the Task.

This is a bit controversial. Tasks are still preferred over value tasks. Micro benchmarking shows a bit of perf cost when using value tasks. They are easier to use correctly. 

According to [this article](https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/#should-every-new-asynchronous-api-return-valuetask-valuetasktresult) a Value task should only be used if:
a) you expect consumers of your API to only await them directly, 
b) allocation-related overhead is important to avoid for your API, and 
c) either you expect synchronous completion to be a very common case, or you’re able to effectively pool objects for use with asynchronous completion. 

I would ask @roji's help/comment on this as well.

Reply:
This fits into cases b and c. The fact that we can't change the external api surface easily to be ValueTask instead of Task based prevents it being a as well.

> This is a bit controversial. Tasks are still preferred over value tasks. Micro benchmarking shows a bit of perf cost when using value tasks. They are easier to use correctly.

One of the things I do for fun is listen to the framework design council meetings when they're streamed on YouTube and I've heard the long conversations about when ValueTask is suitable and when it is not. I investigated this because I had good reason to believe that using ValueTasks is appropriate and would sometimes allow us to have allocation free sync calls. The items I highlighted in the original post show the reduction in allocations from these changes. There will be speed increases as well because we spend less time in GC but I didn't think it necessary to quantify them because they weren't the reason for making the changes.

Note that this works on netcore3 best because the runtime did the work to make SSLStream more efficient by adding ValueTask overloads, we're just following the same pattern to allow the most efficient reads we can.

Reply:
I do agree it's worth introducing ValueTask implementations here - this seems to be a Stream implementation that all driver traffic goes through, and it could indeed reduce lots of allocations when the call completes synchronously (and possibly even asynchronously, depending on how optimized the rewritten SslStream is). There's indeed a lot of discussions around whether to **expose** ValueTask vs. Task when designing APIs, but when **consuming** an API that exposes ValueTask it's generally good practice to use that (unless you're sure the invocation will return asynchronously, and that no asynchronous-completion optimizations are in place in the implementation).

> In my benching run it increases throughput by ~30%.

That seems like quite a high number for this change... FWIW I always find it helpful to post the actual BenchmarkDotNet results and code, to show what exactly is being tested etc.

Reply:
However... @Wraith2 I'm not seeing changes to actually call the ValueTask methods, e.g. SNIPacket.ReadFromStreamAsync still calls the Task-returning overload of Stream.ReadAsync, and chains a ContinueWith call on it. I'm not sure what value that would have - it gets a `ValueTask<int>`, only to convert it to a `Task<int>` (so this should in fact increase overhead).

What am I missing here?

Reply:
I was surprised by the numbers but remember it's throughput not speed. The number comes from looking at the number of Fortunes objects allocated in before and after profiles. Since the profile is limited to a specific time it can be extrapolated. I didn't measure increase more carefully because it wasn't speed I was aiming to improve, it was allocations.

Reply:
Using ValueTask in SNIPacket is on my backlog, i'll get there eventually.

Reply:
I'd be interested in understanding how allocations were reduced if the codebase is still calling the Task-returning overloads though... I'm not doubting what you say - just trying to understand the reduction is working (is there another path I'm missing which calls the ValueTask-returning overloads?). 

> Using ValueTask in SNIPacket is on my backlog, i'll get there eventually.

I'm not sure how you'd do that, considering that ContinueWith works with Task only. And since that is used all across the codebase, I think that moving towards ValueTask would require a more general overhaul towards async/await - which is something we've discussed in the past.

Re this PR specifically... Assuming my understanding is correct and the codebase still calls the Task-returning overloads, then I don't see how there's any value in this change on its own...

Reply:
The change to `ConcurrentQueueSemaphore` is a big part of it, and I think I already split that out into another PR because it was so valuable. It prevents the allocation of a TCS, Task, and two continue objects each time we want to check a lock (which is usually uncontested) for every single packet receipt.

The other changes allow value task paths to be used further down. Without this PR I can't use them in SNIPacket so I've got to start somewhere and I started here. If I'd tried to use them in SNIPacket first you'd complain that it was no use doing that because SslStream didn't have it in. If I try to enable ValueTask everywhere in one PR it'll be a very large PR would be treated with even more suspicion than this one already is.

If you like just ignore the throughput increase and focus on what I was really trying to do here which was enable ValueTask incrementally. At the moment we still have to use AsTask() which clearly isn't a good thing but if the ValueTask paths work and are just converted to Task later then it lets me cleanly start to use ValueTask later for gains.

I used some of the existing runtime stream implementations to inform me on implementing these methods. I can't remember which one exactly because it's been 4 months but I didn't just make all of this up.

Reply:
> The change to ConcurrentQueueSemaphore is a big part of it, and I think I already split that out into another PR because it was so valuable. It prevents the allocation of a TCS, Task, and two continue objects each time we want to check a lock (which is usually uncontested) for every single packet receipt.

It's great that you're making progress. I haven't review this PR thoroughly, though specifically for making ConcurrentQueueSemaphore not allocate for the uncontended case, this seems to [already be the case as far as I can tell](https://github.com/dotnet/SqlClient/pull/902/files#diff-992e2003f7c476cd1e0a3de1867f48ecddbb82ad2a57951db415a8aead4eb92aL2165) seems to have already done that for the uncontested case. Also see [my comment](https://github.com/dotnet/SqlClient/pull/902/files#r653146582) on the non-utility of non-generic ValueTask there.

> The other changes allow value task paths to be used further down. Without this PR I can't use them in SNIPacket so I've got to start somewhere and I started here. If I'd tried to use them in SNIPacket first you'd complain that it was no use doing that because SslStream didn't have it in.

@Wraith2 I'm not complaining about anything. My opinion was asked for, and I'm pointing out that this change, as it currently is, doesn't improve performance in any way, and indeed may slightly regress it. I very much support improving how async is implemented in SqlClient, though I think the real problems really are elsewhere, and not in Task vs. ValueTask.

Reply:
> specifically for making ConcurrentQueueSemaphore not allocate for the uncontended case, this seems to already be the case as far as I can tell

I split that one out after talking with cheena a few months back. It has a big impact and I wanted to make sure it got in as soon as possible. I don't keep rebasing my open PR's until there's some reason to do so, there's no point me taking the time if no-one is ready to review them, I used to do that but it took ages and the number of branches I'd have to review got out hand.

Outisde a microbenchmark designed to show the difference between ValueTask and Task implementations do you think there is likely to be a noticeable performance difference between the versions? I accept and agree with your statements but is it either harmful or wrong to have a ValueTask version?

If you can point me at the other problems I can look at them. Until then I'll keep investigating and addressing the problems I find. I still hope that one day we'll be able to get Sql Server included in the TE benchmarks. The TE Fortunes benchmark is also a simple and easily understood optimization scenario to work on so I'll come back to it often to see if there are new optimization opportunities.

Reply:
> I accept and agree with your statements but is it either harmful or wrong to have a ValueTask version?

No - I don't think it's either harmful or wrong. As I wrote above, I'd implement the Task-returning versions by calling the base SslStream counterparts (rather than the ValueTask), but that's a really small suggestion.

> The TE Fortunes benchmark is also a simple and easily understood optimization scenario to work on so I'll come back to it often to see if there are new optimization opportunities.

I agree that TE Fortunes is a great way to find optimization opportunities (regardless of whether results are officially published on techempower.com or not). But I'll go out on a limb here, and say that IMHO a few micro-optimizations here and there simply aren't going to have a substantial impact given the larger problems in the codebase. At some point these need to be addressed.

PS I'm still unclear on where the aforementioned allocation reductions and throughput improvements come from in this PR, since the new Stream ValueTask methods aren't called, and I can't see how the ConcurrentQueueSemaphore changes improve anything.

Reply:
Ignore the semaphore stuff, that got merged separately. It was the perf improvement.
The stream valuetasks just enable me to use them in a future PR.

Reply:
OK, though consider https://github.com/dotnet/SqlClient/pull/902#discussion_r653146582.